### PR TITLE
Listen to Shiny's idle/busy JS events

### PR DESCRIPTION
**Pull request by schloerke**
To be merged after #8 

* Listen to shiny busy/idle JS events
* Debouce exec of "on idle" functions at 200ms. Prevents thrashing of idle/busy
* Cancel "on idle" call if _busy_

* Reduced exec wait time to 10ms from 250ms

### CODE REVIEW

File: srcjs/methods/shiny.ts [hunk 1]
```diff
@@ -1,28 +1,54 @@
...
-  const wait = function() {
-    if (isBusy()) {
-      setTimeout(wait, timeout);
-    } else {
-      callback();
-    }
-  };
+if ($) {
+  $(document).on("shiny:busy", function(event) {
+    console.log("busy!");
```
Comment (wch) [inline srcjs/methods/shiny.ts]: Are these logging messages needed?

Comment (schloerke) [reply srcjs/methods/shiny.ts]: No. Removing.

File: srcjs/methods/shiny.ts [hunk 2]
```diff
@@ -1,28 +1,51 @@
...
-    }
-  };
+if ($) {
+  $(document).on("shiny:busy", function(event) {
+    shinyIsIdle = false;
+    callIdleFnsDebounced.cancel();
+  });
+  $(document).on("shiny:idle", function(event) {
+    shinyIsIdle = true;
+    // to avoid idle and busy thrashing,
```
Comment (wch) [inline srcjs/methods/shiny.ts]: Can you give a short description of a case where busy/idle thrashing causes problems?

**Comment by schloerke** [comment]
@wch 

Update on the event handlers... Now each `waitUntilIdleFor` (new method) creates unique `shiny:busy`/`shiny:idle` event handlers.  Once the specified timeout time has elapsed, the event handlers are removed and the callback is called.

* No buildup of no-op events
* Users can now supply their own timeout value.

**Pull request merged**